transport/src/main/java/io/servicefabric/transport/protocol/MessageSchema.java

```diff
@@ -135,38 +128,43 @@ public void mergeFrom(Input input, Message message) throws IOException {
     }
 
     // Deserialize headers
+    Map<String, String> headers = new HashMap<>(headerKeys.size());
     if (!headerKeys.isEmpty()) {
-      Map<String, String> headers = new HashMap<>(headerKeys.size());
       ListIterator<String> headerValuesIterator = headerValues.listIterator();
       for (String key : headerKeys) {
         String value = headerValuesIterator.next();
         headers.put(key, value);
       }
-      message.setHeaders(headers);
     }
 
     // Deserialize data
-    if (dataPayload != null) {
-      if (dataType == null) {
-        message.setData(dataPayload);
+    Object data = null;
+    if (dataBytes != null) {
+      String dataClassName = headers.get(TransportHeaders.DATA_CLASS);
+      if (dataClassName == null) {
+        data = dataBytes;
       } else {
-        Optional<Class> optionalDataClass = classCache.get(dataType);
+        Optional<Class> optionalDataClass = classCache.get(dataClassName);
         if (optionalDataClass.isPresent()) {
+          headers.remove(TransportHeaders.DATA_CLASS);
```

Comment:
Since data will be deserialized to java class so we can eliminate this additional header from here.